velox/functions/prestosql/aggregates/ReservoirSampleAggregate.cpp

```diff
@@ -0,0 +1,910 @@
+    vector_size_t size,
+    int64_t seenCount,
+    int32_t desiredSampleSize,
+    memory::MemoryPool* pool) {
+  VELOX_CHECK_NOT_NULL(sampleVector);
+  VELOX_CHECK_GE(seenCount, 0);
+  VELOX_DCHECK(size == seenCount || size == desiredSampleSize);
+
+  if (samples == nullptr) {
+    samples = BaseVector::create(sampleVector->type(), desiredSampleSize, pool);
+  }
+  samples->copy(sampleVector.get(), 0, offset, size);
+  processedCount = seenCount;
+  maxSampleSize = desiredSampleSize;
+}
+
+void ReservoirSampleAccumulator::setInitialSamples(
+    const VectorPtr initialSampleVector,
+    vector_size_t offset,
+    vector_size_t size,
+    int64_t initialProcessedCount,
+    memory::MemoryPool* pool) {
+  if (initialProcessedCount <= 0) {
+    VELOX_USER_CHECK(
+        initialSampleVector == nullptr || size == 0,
+        "Initial state array must be null or empty "
+        "when initial processed count is <= 0.");
+  }
+
+  if (initialSeenCount >= 0) {
+    return;
+  }
+
+  if (initialSampleVector && size > 0) {
+    VELOX_USER_CHECK_GE(
+        initialProcessedCount,
+        size,
+        "InitialProcessedCount must be greater than or equal to the number "
+        "of positions in the initial sample.");
+  }
+
+  // When initialSamples == nullptr, initialSeenCount can be any of
+  // -1, 0, or positive.
+  initialSeenCount = initialProcessedCount;
+
+  if (initialProcessedCount <= 0 || initialSampleVector == nullptr) {
+    initialSamples = nullptr;
+    return;
+  }
+  initialSamples = BaseVector::create(initialSampleVector->type(), size, pool);
+  initialSamples->copy(initialSampleVector.get(), 0, offset, size);
+}
+
+void ReservoirSampleAccumulator::addValueToReservoir(
+    const BaseVector* vector,
+    vector_size_t row,
+    bool isNull,
+    folly::ThreadLocalPRNG& rng) {
+  VELOX_CHECK(isInitialized());
+
+  processedCount = processedCount.value_or(0) + 1;
+
+  vector_size_t targetIndex = -1;
+  if (processedCount.value() <= maxSampleSize) {
+    targetIndex = processedCount.value() - 1;
+  } else {
+    const uint64_t random = folly::Random::rand64(processedCount.value(), rng);
+    if (random < samples->size()) {
+      targetIndex = random;
+    }
+  }
+
+  if (targetIndex >= 0) {
+    if (isNull) {
+      samples->setNull(targetIndex, true);
+    } else {
+      samples->copy(vector, targetIndex, row, 1);
+    }
+  }
+}
+
+void ReservoirSampleAccumulator::addValuesToReservoir(
+    const BaseVector* vector,
+    vector_size_t offset,
+    vector_size_t size,
+    folly::ThreadLocalPRNG& rng) {
+  VELOX_CHECK(isInitialized());
+
+  SelectivityVector rows(maxSampleSize.value(), false);
+  std::vector<vector_size_t> toSource(maxSampleSize.value());
+
+  for (vector_size_t i = 0; i < size; i++) {
+    if (processedCount.value() < maxSampleSize) {
+      toSource[processedCount.value()] = offset + i;
+      rows.setValid(processedCount.value(), true);
+
+      processedCount = processedCount.value_or(0) + 1;
```

Comment:
nit: processedCount ++ is good enough, since in the if condition you already accessed .value() so it should have a value (if it doesnt it will crash there).